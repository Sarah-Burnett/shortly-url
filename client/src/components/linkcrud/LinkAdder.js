import React, { useState } from "react";
import axios from "axios";
import validator from "validator";
import styled from "styled-components";
import { red, cyan, white, breakpoint } from "../Styles";
import Error from "../Error";
import { useUser } from "../../hooks/useUser";
import { useLinks } from "../../hooks/useLinks";
import { addLocalStorage } from "../../utilities/setLocalStorage";

const Form = styled.form`
	margin: auto;
	padding: 1.5vh 1vw;
	@media (max-width: ${breakpoint}) {
		input {
			width: 80vw;
			margin: 2vh 0;
		}
	}
	@media (min-width: ${breakpoint}) {
		width: 80vw;
		display: grid;
		padding: 2vh 2vw 0;
		grid-template-columns: 8fr 2fr;
		grid-template-rows: auto auto;
		grid-column-gap: 1.5vw;
		grid-template-areas: "link button" "error error";
		p {
			margin: 0;
		}
	}
`;

const Input = styled.input`
	display: block;
	border-radius: 5px;
	padding: 1vh 2vw;
	border: none;
	&:invalid {
		border: 2px solid ${red};
	}
	grid-area: link;
`;

const Button = styled(Input)`
	font-weight: bold;
	background: ${cyan};
	color: ${white};
	padding: 1vh 2vw;
	cursor: pointer;
	&:hover {
		color: ${white};
		opacity: 0.5;
	}
	grid-area: button;
`;

const LinkAdder = () => {
	const { user } = useUser();
	const { links, addLink } = useLinks();
	const [input, setInput] = useState({ value: "Shorten a link here... " });
	const [error, setError] = useState(false);
	const [isSending, setIsSending] = useState(false);
	const changeState = (value, valid, err) => {
		setInput({ value: value, valid: valid });
		err ? setError("Please add a link") : setError(false);
	};
	const handleChange = (event) => {
		event.persist();
		setError("");
		const value = event.target.value;
		if (validator.isEmpty(value)) changeState(value, false, true);
		else if (validator.isURL(value)) changeState(value, true, false);
		else changeState(value, false, false);
	};
	const postLink = (event) => {
		event.preventDefault();
		if (!input.valid) return setError("Please add a link");
		const params = { full: input.value };
		if (user.id) params.userId = user.id;
		setIsSending(true);
		axios
			.post("api/links", params)
			.then(({ data }) => {
				const newLinks = [
					...links,
					{
						_id: data[0]._id,
						full: data[0].full,
						short: data[0]._id,
						date: data[0].date,
						click: data[0].click,
					},
				];
				if (!user.id) addLocalStorage("links", JSON.stringify(newLinks));
				addLink(newLinks);
				setInput({ value: "" });
				setIsSending(false);
			})
			.catch(() => {
				setError("Error! Please try again");
				setIsSending(false);
			});
	};
	return (
		<Form onSubmit={(event) => postLink(event)}>
			<label htmlFor="link" style={{ display: "none" }}>
				Website Url
			</label>
			{error && <Error error={error} />}
			<Input
				id="link"
				name="link"
				onChange={handleChange}
				value={input.value}
				required={input.value === "Shorten a link here..." ? false : true}
				onFocus={() => changeState("", false, true)}
				data-testid="input"
			/>
			<Button
				type="submit"
				value={isSending ? "Shortening..." : "Shorten It!"}
				disabled={isSending}
				data-testid="button"
			/>
		</Form>
	);
};

export default LinkAdder;
