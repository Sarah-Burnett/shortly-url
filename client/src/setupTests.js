import React from 'react';
import ReactDOM from 'react-dom';
import '@testing-library/jest-dom/extend-expect';
import { render, cleanup, fireEvent, screen, waitForElement } from '@testing-library/react';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import renderer from 'react-test-renderer';
import { UserProvider } from './context/UserContext';
import { LinksProvider } from './context/LinksContext';
import { ActiveLinkProvider } from './context/ActiveLinkContext';
import { ProcessLinkProvider } from './context/ProcessLinkContext';
import { FlashProvider } from './context/FlashContext';
import App from './App';

export const ComponentWrapper = (Component) => {
    return (
        <UserProvider>
            <LinksProvider>
                <ActiveLinkProvider>
                    <ProcessLinkProvider>
                        <FlashProvider>
                            <Component/>
                        </FlashProvider>
                    </ProcessLinkProvider>
                </ActiveLinkProvider>
            </LinksProvider>
        </UserProvider>
    )
}

export const rendersWithoutCrashing = (Component) => {
    const div = document.createElement('div');
    ReactDOM.render(<Component/>, div);
}

export const matchesSnapshot = (Component) => {
    const tree = renderer.create(<Component/>).toJSON();
    expect(tree).toMatchSnapshot();
}

export const sampleLinks = [
    {
        "date":"2020-08-14T17:29:53.041Z",
        "click":[
            {"date":1597426302926},
            {"date":1597426306071},
            {"date":1597426310231},
            {"date":1597426314763},
            {"date":1597428206995},
            {"date":1597428210093},
            {"date":1597428215131},
            {"date":1597428218755}
        ],
        "_id":"K1pjMCK-F",
        "full":"http://reactcommunity.org/react-transition-group/css-transition",
        "userId":"5ebe7d235e4a501ee8699058",
        "short":"octopus","__v":8
    },
    {
        "date":"2020-08-14T17:33:55.748Z",
        "click":[],
        "_id":"lsncCe5RV",
        "full":"https://ba-grant.netlify.app/",
        "userId":"5ebe7d235e4a501ee8699058",
        "short":"bagrant","__v":0
    }
];
