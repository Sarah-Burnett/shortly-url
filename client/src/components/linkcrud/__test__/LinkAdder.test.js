import { rendersWithoutCrashing, matchesSnapshot, ComponentWrapper } from '../../../setupTests';
import { render, fireEvent, screen } from '@testing-library/react';
import LinkAdder from '../LinkAdder';

const LinkAdderWrapper = () => ComponentWrapper(LinkAdder);

it('renders link adder without crashing', () => rendersWithoutCrashing(LinkAdderWrapper));

it('link adder matches snapshot', () => matchesSnapshot(LinkAdderWrapper));

it('valid url does not show error', async () => {
    render(LinkAdderWrapper());
    await fireEvent.change(screen.getByTestId('input'), {
        target: {
            value: "https://google.com"
        }
    })
    await fireEvent.click(screen.getByTestId('button'))
    expect(screen.queryByTestId('error')).toBe(null);
})

it('invalid url shows error', async () => {
    render(LinkAdderWrapper());
    await fireEvent.change(screen.getByTestId('input'), {
        target: {
            value: "asdfjalsfjwpjfroejf"
        }
    })
    await fireEvent.click(screen.getByTestId('button'))
    expect(screen.getByTestId('error')).toHaveTextContent('Please add a link');
})

//TODO: check adding link adds link - integration