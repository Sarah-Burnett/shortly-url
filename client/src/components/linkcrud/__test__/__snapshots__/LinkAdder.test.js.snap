// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`link adder matches snapshot 1`] = `
<form
  className="sc-fznyAO dlGfaB"
  onSubmit={[Function]}
>
  <label
    htmlFor="link"
    style={
      Object {
        "display": "none",
      }
    }
  >
    Website Url
  </label>
  <p
    className="sc-fzokOt kPqHuq"
  />
  <input
    className="sc-fznKkj jhFdRJ"
    data-testid="input"
    id="link"
    name="link"
    onChange={[Function]}
    onFocus={[Function]}
    required={true}
    value="Shorten a link here... "
  />
  <input
    className="sc-fznKkj sc-fznZeY iOAXEI"
    data-testid="button"
    disabled={false}
    type="submit"
    value="Shorten It!"
  />
</form>
`;
